src/plugins/discover/public/application/main/hooks/use_discover_state.ts

```diff
@@ -220,12 +245,53 @@ export function useDiscoverState({
   /**
    * Trigger data fetching on indexPattern or savedSearch changes
    */
+  useEffect(() => {
+    if (!isEqual(state.query, prevQuery)) {
+      setDocumentStateCols([]);
+    }
+  }, [state.query, prevQuery]);
+
   useEffect(() => {
     if (indexPattern) {
       refetch$.next(undefined);
     }
   }, [initialFetchStatus, refetch$, indexPattern, savedSearch.id]);
 
+  const getResultColumns = useCallback(() => {
+    if (documentState.result?.length && documentState.fetchStatus === FetchStatus.COMPLETE) {
+      const firstRow = documentState.result[0];
+      const columns = Object.keys(firstRow.raw).slice(0, MAX_NUM_OF_COLUMNS);
+      if (!isEqual(columns, documentStateCols) && !isEqual(state.query, sqlQuery)) {
+        return columns;
+      }
+      return [];
+    }
+    return [];
+  }, [documentState, documentStateCols, sqlQuery, state.query]);
+
+  useEffect(() => {
+    async function fetchDataview() {
+      if (state.query && isOfAggregateQueryType(state.query) && 'sql' in state.query) {
+        const indexPatternFromQuery = getIndexPatternFromSQLQuery(state.query.sql);
+        const idsTitles = await indexPatterns.getIdsWithTitle();
+        const dataViewObj = idsTitles.find(({ title }) => title === indexPatternFromQuery);
+        if (dataViewObj) {
+          const columns = getResultColumns();
+          if (columns.length) {
+            setDocumentStateCols(columns);
+          }
+          const nextState = {
+            index: dataViewObj.id,
+            ...(columns.length && { columns }),
+          };
+          stateContainer.setAppState(nextState);
```

Comment:
I think this should solve the back button problem @davismcphee mentioned.

So we're setting the columns after the SQL query is executed. And then we modify the app state, which pushes to history. With the following change history is replace, not pushed:

```suggestion
          stateContainer.replaceUrlAppState(nextState);
```

Reply:
sorry for this mess here, can't wait to refactor 😄